### DISCUSSION TITLE
How do I do programmatic construction of SdkConfig?

### DISCUSSION THREAD

**Discussion by sazzer** [awslabs/aws-sdk-rust#672] [Q&A] [answered]
I'm writing a non-AWS deployed service that is using AWS features - Cognito, specifically. And I'm wanting to be able to programmatically construct the `SdkConfig` object, mostly so that I can better control how the values are loaded and the errors that are thrown when they aren't available. As it is right now, when loading an `SdkConfig` using the standard way, if there are no credentials provided then it all works fine until you come to actually use it!

What I've done so far is:
```
        let aws_region = "eu-west-1".to_owned();

        let aws_config = aws_config::ConfigLoader::default()
            .region(
                aws_config::default_provider::region::Builder::default()
                    .profile_name(&aws_region)
                    .build(),
            )
            .load()
            .await;
```

Where `aws_region` is a field that, for now, is hard-coded but is going to be loaded from my configuration in time. This seems to work fine, right up until you run the program with the AWS_REGION environment variable set. As soon as you do that, the configured value is ignored in favour of that environment variable. :(

Is there some correct way to achieve this? Or am I just wasting my time?

Cheers

**Comment by jdisanti** [comment, marked as answer]
[Environment variables take priority over profile files](https://docs.rs/aws-config/0.51.0/aws_config/default_provider/region/fn.default_provider.html) in the default provider. If you want to force it to use the profile file instead of the env var, you can supply it with a [`ProfileFileRegionProvider`](https://docs.rs/aws-config/0.51.0/aws_config/profile/struct.ProfileFileRegionProvider.html) directly instead of `aws_config::default_provider::region::Builder...`.

**Comment by sazzer** [comment]
But is there a decent way for me to provide the region and credentials that
I've loaded myself? Or else is there a way to have the program terminate
with a decent error message when loading the config? Because right now what
happens is that everything works until the first time the API is called,
and then that call returns an error when it would be much nicer to have
service startup fail instead ..

Cheers

**Reply by jdisanti** [reply]
The credential provider is only run upon request by design since a lot can change in between initialization and first request (for example, credentials can expire before the first request is made if they're eagerly retrieved, or the program itself could change the profile file before making a request). If you want a startup check, you can grab [the provider out of the resolved `SdkConfig`](https://docs.rs/aws-config/0.51.0/aws_config/struct.SdkConfig.html#method.credentials_provider) and call the [`provide_credentials`](https://docs.rs/aws-types/0.51.0/aws_types/credentials/trait.ProvideCredentials.html#tymethod.provide_credentials), which will return a `Result` (after `await`) that you can check to see if everything is working correctly.

The region provider gets called when resolving the initial `SdkConfig` since regions can't be changed after initialization, so you can just check the `Option` that the [`region`](https://docs.rs/aws-config/0.51.0/aws_config/struct.SdkConfig.html#method.region) function returns.

**Reply by sazzer** [reply]
Ok, so I've got this and it seems to work well enough - in that if I don't specify all of the environment properties my startup fails with an error message telling me roughly enough what's happened:

```
        let aws_config = aws_config::ConfigLoader::default().load().await;

        let aws_region = aws_config
            .region()
            .expect("No AWS region was provided")
            .to_string();

        aws_config
            .credentials_provider()
            .expect("No AWS credentials provider was configured")
            .provide_credentials()
            .await
            .expect("No AWS credentials were provided");
```

```
$ cargo run
...
thread 'main' panicked at 'No AWS credentials were provided: CredentialsNotLoaded { context: "no providers in chain provided credentials" }', src/main.rs:77:14
```

It'd be nice if it told me exactly what was missing - in this case there was no `AWS_SECRET_ACCESS_KEY` environment variable defined - but it'll do :)
